File: VB/WinDashboard_LinkedInteractivity/Form2.vb
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardCommon.ViewerData
Imports DevExpress.DashboardWin
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq

Namespace WinDashboard_LinkedInteractivity
	Partial Public Class Form2
		Inherits DevExpress.XtraEditors.XtraForm

		Public Shadows Property ParentForm() As Form1

		Public Sub New(ByVal parentForm As Form1)
			InitializeComponent()
			Me.dashboardViewer1.LoadDashboard("Data\DashboardChild.xml")
			Me.ParentForm = parentForm
			Dim dParentControl As DashboardDesigner = Me.ParentForm.DControl
			AddHandler dParentControl.MasterFilterSet, AddressOf DControl_MasterFilterSet
			AddHandler dParentControl.MasterFilterCleared, AddressOf DControl_MasterFilterCleared
			AddHandler dParentControl.DrillDownPerformed, AddressOf DControl_DrillDownPerformed
			AddHandler dParentControl.DrillUpPerformed, AddressOf DControl_DrillUpPerformed
		End Sub

		Private Function HasDashboardItem(ByVal itemName As String) As Boolean
			Return Me.dashboardViewer1.Dashboard.Items.Select(Function(i) i.ComponentName).Contains(itemName)
		End Function

		Private Sub DControl_MasterFilterSet(ByVal sender As Object, ByVal e As MasterFilterSetEventArgs)
			If HasDashboardItem(e.DashboardItemName) Then
				Dim itemName As String = e.DashboardItemName
				If e.SelectedValues IsNot Nothing Then
					Me.dashboardViewer1.SetMasterFilter(itemName, e.SelectedValues)
				End If
				If e.SelectedRange IsNot Nothing Then
					Me.dashboardViewer1.SetRange(itemName, e.SelectedRange)
				End If
			End If
		End Sub
		Private Sub DControl_MasterFilterCleared(ByVal sender As Object, ByVal e As MasterFilterClearedEventArgs)
			If HasDashboardItem(e.DashboardItemName) Then
				Dim itemName As String = e.DashboardItemName
				If Me.dashboardViewer1.CanClearMasterFilter(itemName) Then
					Me.dashboardViewer1.ClearMasterFilter(itemName)
				End If
			End If
		End Sub
		Private Sub DControl_DrillDownPerformed(ByVal sender As Object, ByVal e As DrillActionEventArgs)
			If HasDashboardItem(e.DashboardItemName) Then
				Dim itemName As String = e.DashboardItemName
				Dim row As DashboardDataRow = e.Values(0)
				Dim value As Object = row(row.Length - 1)

				Dim tuple As IList(Of AxisPointTuple) = Me.dashboardViewer1.GetAvailableDrillDownValues(itemName)
				Dim availableValues As IEnumerable(Of Object) = tuple.Select(Function(t) t.GetAxisPoint().UniqueValue)
				If availableValues.Contains(value) Then
					Me.dashboardViewer1.PerformDrillDown(e.DashboardItemName, value)
				End If
			End If
		End Sub
		Private Sub DControl_DrillUpPerformed(ByVal sender As Object, ByVal e As DrillActionEventArgs)
			If HasDashboardItem(e.DashboardItemName) Then
				Dim itemName As String = e.DashboardItemName
				Dim level As Integer = e.DrillDownLevel
				Dim tuple As AxisPointTuple = Me.dashboardViewer1.GetCurrentDrillDownValues(itemName)
				If tuple IsNot Nothing Then
					Dim point As AxisPoint = Me.dashboardViewer1.GetCurrentDrillDownValues(itemName).GetAxisPoint()
					Dim l As Integer = 0
					Do While point.Parent IsNot Nothing
						l += 1
						point = point.Parent
					Loop
					If level + 1 = l AndAlso Me.dashboardViewer1.CanPerformDrillUp(itemName) Then
						Me.dashboardViewer1.PerformDrillUp(itemName)
					End If
				End If
			End If
		End Sub
	End Class
End Namespace

File: CS/WinDashboard_LinkedInteractivity/Form2.cs
using DevExpress.DashboardCommon;
using DevExpress.DashboardCommon.ViewerData;
using DevExpress.DashboardWin;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace WinDashboard_LinkedInteractivity {
    public partial class Form2 : DevExpress.XtraEditors.XtraForm {
        public new Form1 ParentForm { get; set; }        
        public Form2(Form1 parentForm) {
            InitializeComponent();
            this.dashboardViewer1.LoadDashboard("Data\\DashboardChild.xml");
            this.ParentForm = parentForm;
            DashboardDesigner dParentControl = this.ParentForm.DControl;
            dParentControl.MasterFilterSet += DControl_MasterFilterSet;
            dParentControl.MasterFilterCleared += DControl_MasterFilterCleared;
            dParentControl.DrillDownPerformed += DControl_DrillDownPerformed;
            dParentControl.DrillUpPerformed += DControl_DrillUpPerformed;
        }

        bool HasDashboardItem(string itemName) {
            return this.dashboardViewer1.Dashboard.Items.
                Select(i => i.ComponentName).Contains(itemName);
        }

        private void DControl_MasterFilterSet(object sender, MasterFilterSetEventArgs e) {
            if (HasDashboardItem(e.DashboardItemName)) {
                string itemName = e.DashboardItemName;
                if (e.SelectedValues != null)
                    this.dashboardViewer1.SetMasterFilter(itemName, e.SelectedValues);
                if (e.SelectedRange != null)
                    this.dashboardViewer1.SetRange(itemName, e.SelectedRange);
            }
        }
        private void DControl_MasterFilterCleared(object sender, MasterFilterClearedEventArgs e) {
            if (HasDashboardItem(e.DashboardItemName)) {
                string itemName = e.DashboardItemName;
                if (this.dashboardViewer1.CanClearMasterFilter(itemName))
                    this.dashboardViewer1.ClearMasterFilter(itemName);
            }
        }
        private void DControl_DrillDownPerformed(object sender, DrillActionEventArgs e) {
            if (HasDashboardItem(e.DashboardItemName)) {
                string itemName = e.DashboardItemName;
                DashboardDataRow row = e.Values[0];
                object value = row[row.Length - 1];

                IList<AxisPointTuple> tuple = this.dashboardViewer1.GetAvailableDrillDownValues(itemName);
                IEnumerable<object> availableValues = tuple.Select(t => t.GetAxisPoint().UniqueValue);
                if (availableValues.Contains(value)) {
                    this.dashboardViewer1.PerformDrillDown(e.DashboardItemName, value);
                }
            }
        }
        private void DControl_DrillUpPerformed(object sender, DrillActionEventArgs e) {
            if (HasDashboardItem(e.DashboardItemName)) {
                string itemName = e.DashboardItemName;
                int level = e.DrillDownLevel;
                AxisPointTuple tuple = this.dashboardViewer1.GetCurrentDrillDownValues(itemName);
                if (tuple != null) {
                    AxisPoint point = this.dashboardViewer1.GetCurrentDrillDownValues(itemName).GetAxisPoint();
                    int l = 0;
                    while (point.Parent != null) {
                        l++;
                        point = point.Parent;
                    }
                    if (level + 1 == l && this.dashboardViewer1.CanPerformDrillUp(itemName))
                        this.dashboardViewer1.PerformDrillUp(itemName);
                }
            }
        }
    }
}
